lib/matplotlib/figure.py

```diff
@@ -1582,10 +1582,10 @@ def subfigures(self, nrows=1, ncols=1, squeeze=True,
                       height_ratios=height_ratios,
                       left=0, right=1, bottom=0, top=1)
 
-        sfarr = np.empty((nrows, ncols), dtype=object)
-        for i in range(ncols):
-            for j in range(nrows):
-                sfarr[j, i] = self.add_subfigure(gs[j, i], **kwargs)
+        sfarr = np.empty((ncols, nrows), dtype=object)
```

Comment:
I think the order in the array was correct:  if we adapt @anntzer's example to use the returned array instead of the list attribute then the order is consistent

```python
from pylab import *

f = figure()
ax_arr = f.subplots(3, 3)
for i, ax in enumerate(ax_arr.ravel()): ax.set_title(i)
savefig("/tmp/subplots.png")

f = figure()
sf_arr = f.subfigures(3, 3)
for i, sf in enumerate(sf_arr.ravel()): sf.suptitle(i)
savefig("/tmp/subfigures.png")
```
![subplots](https://github.com/matplotlib/matplotlib/assets/10599679/78e3eece-0e6a-4157-a054-9349153f0270)
![subfigures](https://github.com/matplotlib/matplotlib/assets/10599679/f99a1b04-0493-45cd-b4a6-b8fa886a30ac)

So I think we just need to keep the array the same but just swap the nesting of the if-loops.